### ISSUE TITLE
cuda.sparse_spmv_tensor_core failures with clang/7+cuda/9

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos-kernels#1314]
@cwpearson when testing on kokkos-dev-2 after merge of #1307 I encountered test failures with spmv tensor code tests, can you investigate? I'm not sure if the failures are related to #1307 or showed up with separate changes 

**Output failure snip**:
```
4: [ RUN      ] cuda.sparse_spmv_tensor_core_double_double_double_int_size_t_LayoutLeft_TestExecSpace
4: KokkosSparse::Test::spmv_tc: 16 errors of 16 for mv 0 (alpha=1, beta=0, mode = N)
4: /ascldap/users/ndellin/kokkos-kernels/unit_test/sparse/Test_Sparse_spmv.hpp:1306: Failure
4: Value of: num_errors == 0
4:   Actual: false
4: Expected: true
4: KokkosSparse::Test::spmv_tc: 16 errors of 16 for mv 1 (alpha=1, beta=0, mode = N)
4: /ascldap/users/ndellin/kokkos-kernels/unit_test/sparse/Test_Sparse_spmv.hpp:1306: Failure
4: Value of: num_errors == 0
4:   Actual: false
4: Expected: true
...
```

**Reproducer** (kokkos-dev-2):
```
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/7.2.0 sems-archive-cuda/9.2 sems-archive-clang/7.0.1

$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Volta70 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft
```

**Comment by cwpearson** [comment]
First run
```
source /projects/sems/modulefiles/utils/sems-archive-modules-init.sh
```

**Comment by cwpearson** [comment]
I think this happens because here:

https://github.com/kokkos/kokkos-kernels/blob/2199f9914e4c1e620641b5bc4642c765f70cf7f9/src/Kokkos_ArithTraits.hpp#L655

Kokkos::ArithTraits for half is only properly defined when `KOKKOS_HALF_T_IS_FLOAT = 0`, otherwise, it uses float. This causes the half-precision tolerances to be too tight

**Comment by cwpearson** [comment]
Yes, prec() for half is actually `float` in CUDA 9 on the host.

**Comment by cwpearson** [comment]
And this may be the cause:

https://github.com/kokkos/kokkos/blob/542073765dcdb95a630fefe2547e5e309c7f3f3c/core/src/Cuda/Kokkos_Cuda_Half_Impl_Type.hpp#L50-L53

**Comment by cwpearson** [comment]
Hopefully fixed in #1322

**Comment by cwpearson** [comment]
A more comprehensive fix for the root issue is part of #1329
